File: bench/11_par/bench.py
"""Benchmark 11: par equivalent — 100,000 elements of 50-step Newton's sqrt, 4 processes.

The timed loop runs `--n` ops, each on an input that differs by the repetition number in
VALUE, never in the amount of work, and folds every result into `sink`.  The `hash`
column is one canonical op's result; every lane must print the same one.
"""
import sys
import time

import math
from multiprocessing import Pool


def newton_sqrt(x):
    guess = x / 2.0
    for _ in range(50):
        guess = (guess + x / guess) / 2.0
    return guess


def chunk(bounds):
    lo, hi = bounds
    return [newton_sqrt(float(i + 1)) for i in range(lo, hi)]


def one_pass(pool, count):
    part = count // 4
    bounds = [(t * part, count if t == 3 else (t + 1) * part) for t in range(4)]
    total = 0.0
    for values in pool.map(chunk, bounds):  # summed in item order, as loft's `par` delivers
        for v in values:
            total += v
    return int(math.floor(total + 0.5))

def arg_n(dflt):
    """`--n N`: how many ops the timed loop runs (the harness calibrates it per lane)."""
    n = dflt
    for i, a in enumerate(sys.argv):
        if a == "--n" and i + 1 < len(sys.argv):
            n = int(sys.argv[i + 1])
    return max(n, 2)


def row(name, iters, us, items, result, sink):
    """One measured routine, in the row format every lane prints (bench/README.md)."""
    print("routine\titers\tus\tns_op\tpx\tns_px\thash")
    print(f"{name}\t{iters}\t{us}\t{us * 1000 // iters}\t{items}\t{us * 1000 / (iters * items):.3f}\t{result:x}")
    print(f"time: {us // 1000}ms sink={sink}")


def timed(n, op):
    t0 = time.perf_counter_ns()
    sink = 0
    for r in range(n):
        sink += op(r)
    return (time.perf_counter_ns() - t0) // 1000, sink

if __name__ == "__main__":
    n = arg_n(2)
    count = 100_000
    with Pool(4) as pool:
        warm = one_pass(pool, count)
        us, sink = timed(n, lambda r: one_pass(pool, count))
    row("par", n, us, count, warm, sink)

File: bench/11_par/bench.rs
// Benchmark 11: par equivalent — 100,000 elements of 50-step Newton's sqrt, 4 threads
//
// One OP is one parallel pass: a 4-way range partition, each worker summing its part, the
// main thread adding the parts in order — the shape of loft's `par(r = work(s), 4)`.  No
// external crates: the harness compiles each bench.rs with bare `rustc`.
use std::hint::black_box;
use std::thread;
use std::time::Instant;

fn newton_sqrt(x: f64) -> f64 {
    let mut g = x / 2.0;
    for _ in 0..50 {
        g = (g + x / g) / 2.0;
    }
    g
}

fn chunk(lo: i64, hi: i64) -> Vec<f64> {
    (lo..hi).map(|i| newton_sqrt((i + 1) as f64)).collect()
}

/// One pass.  The workers return their elements and the main thread sums them in item
/// order, because that is the order loft's `par` delivers results in: a float sum taken
/// per worker and then combined would round differently and the hash would not agree.
fn pass(count: i64) -> i64 {
    let workers: i64 = 4;
    let part = count / workers;
    let handles: Vec<_> = (0..workers)
        .map(|t| {
            let lo = t * part;
            let hi = if t == workers - 1 { count } else { lo + part };
            thread::spawn(move || chunk(lo, hi))
        })
        .collect();
    let mut sum = 0.0f64;
    for h in handles {
        for v in h.join().unwrap() {
            sum += v;
        }
    }
    sum.round() as i64
}

fn arg_n(dflt: i64) -> i64 {
    let args: Vec<String> = std::env::args().collect();
    let mut n = dflt;
    let mut i = 1;
    while i < args.len() {
        if args[i] == "--n" && i + 1 < args.len() {
            n = args[i + 1].parse().unwrap_or(dflt);
            i += 1;
        }
        i += 1;
    }
    n.max(2)
}

fn main() {
    let n = arg_n(20);
    let count: i64 = 100_000;
    let warm = pass(count);
    let t0 = Instant::now();
    let mut sink: i64 = 0;
    for _ in 0..n {
        sink = sink.wrapping_add(pass(black_box(count)));
    }
    let us = t0.elapsed().as_micros() as i64;
    let ns = us * 1000 / n;
    let per = (us * 1000) as f64 / (n * count) as f64;
    println!("routine\titers\tus\tns_op\tpx\tns_px\thash");
    println!("par\t{n}\t{us}\t{ns}\t{count}\t{per:.3}\t{warm:x}");
    println!("time: {}ms sink={}", us / 1000, black_box(sink));
}
